File: eticaretfronted/src/types/product.ts
export interface Feature {
  color?: string;
  material?: string;
  size?: string;
}

export interface ProductViewDTO {
  id: number;               // Ürünün benzersiz ID'si
  name: string;             // Ürünün adı
  description: string;      // Ürünün açıklaması
  price: number;            // Ürünün fiyatı
  image_url: string;        // Ürünün ana görsel URL'si
  images?: string[];        // Ürünün diğer görselleri
  categoryName?: string;    // Ürünün kategori adı
  categoryId?: number;      // Ürünün kategori Id
  sku?: string;             // Ürünün stok kodu (SKU)
  stock?: number;           // Ürün stoğu miktarı   
  features?: Feature[];     // Ürünün özellikleri (renk, malzeme, boyut vb.) 
  avgRating?: number;       // Ürünün ortalama puanı
  reviewCount?: number;     // Ürüne yapılan toplam yorum sayısı
  totalSold?: number;       // Ürünün toplam satışı
  createdAt?: string;       // Ürünün eklenme tarihi
  updatedAt?: string;       // Ürünün güncellenme tarihi
}

File: eticaretbackend/eticaretbackend/src/main/java/com/example/eticaretbackend/model/Product.java
package com.example.eticaretbackend.model;

import jakarta.persistence.*;
import jakarta.validation.constraints.Min;
import jakarta.validation.constraints.Positive;
import lombok.*;

import java.util.List;

@Entity
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
public class Product extends BaseEntity {

    // Ürün adı, boş bırakılamaz
    @Column(nullable = false)
    private String name;

    // Ürün kodu (SKU), benzersiz ve boş bırakılamaz
    @Column(unique = true, nullable = false)
    private String sku;

    // Ürün açıklaması, uzun metin olabilir
    @Column(length = 2000)
    private String description;

    // Ürün fiyatı, pozitif olmalı
    @Column(nullable = false)
    @Positive(message = "Price must be positive")
    private Double price;

    // Stok adedi, 0 veya daha fazla olabilir
    @Min(0)
    private Integer stock;

    // Ürün görsel URL'si (opsiyonel)
    @Column(name = "image_url")
    private String imageUrl;

    // Ürün özellikleri JSON formatında saklanabilir (opsiyonel)
    @Column(columnDefinition = "json")
    private String features;


    // Her ürünün bir kategoriye ait olduğunu belirtir ve kategori bilgisi yalnızca ihtiyaç duyulduğunda veritabanından yüklenir.
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "category_id", nullable = false)
    private Category category;

    // Her ürünün birden fazla yorum/incelemesi olabileceğini belirtir ve ürün üzerinde yapılan tüm işlemler ilişkili yorumlara da otomatik uygulanır.
    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL, fetch = FetchType.LAZY)
    private List<ProductReview> reviews;


}
